File: backend/src/main/java/gov/mha/screening/validation/Mrz.java
package gov.mha.screening.validation;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

/**
 * ICAO 9303 TD3 (passport) MRZ parser + check-digit validation.
 * Deterministic, explainable — this is the most reliable fraud signal
 * for passport-number manipulation (Spec Part 1, case #5).
 */
public final class Mrz {

    private static final String WEIGHTS = "731";

    private Mrz() {}

    public record Parsed(
            String documentType,
            String issuingCountry,
            String surname,
            String givenNames,
            String documentNumber,
            String nationality,
            LocalDate dateOfBirth,
            String sex,
            LocalDate expiryDate,
            boolean documentNumberValid,
            boolean dobValid,
            boolean expiryValid,
            boolean finalCheckValid,
            boolean allChecksValid
    ) {}

    public static int charValue(char c) {
        if (c == '<') return 0;
        if (c >= '0' && c <= '9') return c - '0';
        if (c >= 'A' && c <= 'Z') return c - 'A' + 10;
        return -1; // invalid
    }

    public static int checkDigit(String field) {
        int sum = 0;
        for (int i = 0; i < field.length(); i++) {
            int v = charValue(field.charAt(i));
            if (v < 0) return -1;
            sum += v * (WEIGHTS.charAt(i % 3) - '0');
        }
        return sum % 10;
    }

    public static boolean verify(String field, char providedCheck) {
        int computed = checkDigit(field);
        int provided = (providedCheck == '<') ? 0 : Character.digit(providedCheck, 10);
        return computed >= 0 && computed == provided;
    }

    /** Parse a 2-line, 44-char TD3 MRZ. Returns empty if the shape is wrong. */
    public static Optional<Parsed> parseTd3(String raw) {
        if (raw == null) return Optional.empty();
        String[] lines = raw.trim().replace(" ", "").split("\\R");
        if (lines.length < 2) return Optional.empty();
        String l1 = pad(lines[0]);
        String l2 = pad(lines[1]);
        if (l1.length() < 44 || l2.length() < 44) return Optional.empty();

        String documentType = l1.substring(0, 2).replace("<", "");
        String issuingCountry = l1.substring(2, 5).replace("<", "");
        String names = l1.substring(5, 44);
        String surname;
        String given;
        int sep = names.indexOf("<<");
        if (sep >= 0) {
            surname = names.substring(0, sep).replace("<", " ").trim();
            given = names.substring(sep + 2).replace("<", " ").trim();
        } else {
            surname = names.replace("<", " ").trim();
            given = "";
        }

        String docNum = l2.substring(0, 9);
        char docNumCheck = l2.charAt(9);
        String nationality = l2.substring(10, 13).replace("<", "");
        String dob = l2.substring(13, 19);
        char dobCheck = l2.charAt(19);
        String sex = String.valueOf(l2.charAt(20));
        String expiry = l2.substring(21, 27);
        char expiryCheck = l2.charAt(27);
        String optional = l2.substring(28, 42);
        char optionalCheck = l2.charAt(42);
        char finalCheck = l2.charAt(43);

        boolean docNumValid = verify(docNum, docNumCheck);
        boolean dobValid = verify(dob, dobCheck);
        boolean expiryValid = verify(expiry, expiryCheck);

        String composite = docNum + docNumCheck + dob + dobCheck + expiry + expiryCheck + optional + optionalCheck;
        boolean finalValid = verify(composite, finalCheck);

        return Optional.of(new Parsed(
                documentType, issuingCountry, surname, given,
                docNum.replace("<", ""), nationality,
                parseYymmdd(dob, true), sex, parseYymmdd(expiry, false),
                docNumValid, dobValid, expiryValid, finalValid,
                docNumValid && dobValid && expiryValid && finalValid));
    }

    private static String pad(String s) {
        StringBuilder sb = new StringBuilder(s);
        while (sb.length() < 44) sb.append('<');
        return sb.toString();
    }

    private static LocalDate parseYymmdd(String s, boolean isDob) {
        try {
            int yy = Integer.parseInt(s.substring(0, 2));
            int mm = Integer.parseInt(s.substring(2, 4));
            int dd = Integer.parseInt(s.substring(4, 6));
            int nowYy = LocalDate.now().getYear() % 100;
            int century = isDob ? (yy > nowYy ? 1900 : 2000) : (yy < 50 ? 2000 : 1900) + 0;
            if (!isDob) century = 2000; // expiry always assumed 21st century for demo
            return LocalDate.of(century + yy, mm, dd);
        } catch (Exception e) {
            return null;
        }
    }

    @SuppressWarnings("unused")
    private static final DateTimeFormatter YYMMDD = DateTimeFormatter.ofPattern("yyMMdd");
}

File: SIH2026/ai-service/app/services/mrz.py
"""ICAO 9303 TD3 (passport) MRZ parsing and check-digit validation."""
from __future__ import annotations

import re
from dataclasses import dataclass, field

_WEIGHTS = (7, 3, 1)


def char_value(c: str) -> int:
    if c == "<":
        return 0
    if c.isdigit():
        return int(c)
    if "A" <= c <= "Z":
        return ord(c) - ord("A") + 10
    return -1


def check_digit(fieldval: str) -> int:
    total = 0
    for i, c in enumerate(fieldval):
        v = char_value(c)
        if v < 0:
            return -1
        total += v * _WEIGHTS[i % 3]
    return total % 10


def _verify(fieldval: str, provided: str) -> bool:
    computed = check_digit(fieldval)
    prov = 0 if provided == "<" else (int(provided) if provided.isdigit() else -1)
    return computed >= 0 and computed == prov


@dataclass
class ParsedMrz:
    document_type: str = ""
    issuing_country: str = ""
    surname: str = ""
    given_names: str = ""
    document_number: str = ""
    nationality: str = ""
    date_of_birth: str = ""     # ISO yyyy-mm-dd (best effort)
    sex: str = ""
    expiry_date: str = ""
    checks: dict[str, bool] = field(default_factory=dict)

    @property
    def all_valid(self) -> bool:
        return bool(self.checks) and all(self.checks.values())


def _iso_date(yymmdd: str, is_dob: bool) -> str:
    if not re.fullmatch(r"\d{6}", yymmdd):
        return ""
    yy, mm, dd = int(yymmdd[:2]), int(yymmdd[2:4]), int(yymmdd[4:6])
    if is_dob:
        century = 1900 if yy > 30 else 2000
    else:
        century = 2000
    try:
        return f"{century + yy:04d}-{mm:02d}-{dd:02d}"
    except ValueError:
        return ""


def parse_td3(raw: str | None) -> ParsedMrz | None:
    if not raw:
        return None
    lines = [ln.strip().replace(" ", "").upper() for ln in raw.strip().splitlines() if ln.strip()]
    if len(lines) < 2:
        return None
    # Clean lines and normalize fillers
    cleaned_lines = []
    for ln in lines:
        cleaned = re.sub(r"[^A-Z0-9<]", "<", ln)
        cleaned_lines.append(cleaned)
    lines = cleaned_lines

    # Must look like an ICAO 9303 TD3 MRZ:
    # Line 1 must begin with P (passport) and contain '<'
    if not (lines[0].startswith("P") and "<" in lines[0]):
        return None
    # Line 2 must have at least one '<' or digits in DOB/expiry positions
    if lines[1].count("<") < 1 and not any(c.isdigit() for c in lines[1]):
        return None
    l1 = (lines[0] + "<" * 44)[:44]
    l2 = (lines[1] + "<" * 44)[:44]

    # Character confusion repair in Line 2:
    # Digits are strictly expected in DOB (13..19), dob_chk (19), expiry (21..27), expiry_chk (27), optional_chk (42), final_chk (43)
    l2_chars = list(l2)
    digit_fix = {
        "O": "0", "o": "0", "Q": "0", "D": "0",
        "I": "1", "i": "1", "l": "1", "|": "1",
        "Z": "2", "z": "2",
        "S": "5", "s": "5",
        "B": "8",
    }
    for idx in [9, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 25, 26, 27, 42, 43]:
        if idx < len(l2_chars) and l2_chars[idx] in digit_fix:
            l2_chars[idx] = digit_fix[l2_chars[idx]]

    # Alpha characters expected in Nationality (10..13) and Sex (20)
    alpha_fix = {"0": "O", "1": "I", "5": "S", "2": "Z", "8": "B"}
    for idx in [10, 11, 12, 20]:
        if idx < len(l2_chars) and l2_chars[idx] in alpha_fix:
            l2_chars[idx] = alpha_fix[l2_chars[idx]]

    l2 = "".join(l2_chars)

    names = l1[5:44]
    if "<<" in names:
        surname, _, given = names.partition("<<")
    else:
        surname, given = names, ""
    surname = surname.replace("<", " ").strip()
    given = given.replace("<", " ").strip()

    doc_num = l2[0:9]
    doc_num_chk = l2[9]
    nationality = l2[10:13].replace("<", "")
    dob = l2[13:19]
    dob_chk = l2[19]
    sex = l2[20]
    expiry = l2[21:27]
    expiry_chk = l2[27]
    optional = l2[28:42]
    optional_chk = l2[42]
    final_chk = l2[43]

    doc_ok = _verify(doc_num, doc_num_chk)
    dob_ok = _verify(dob, dob_chk)
    exp_ok = _verify(expiry, expiry_chk)

    # ICAO 9303 TD3 composite check digit evaluation:
    # Format A: with optional check digit included (pos 0..10, 13..20, 21..28, 28..43)
    comp_a = doc_num + doc_num_chk + dob + dob_chk + expiry + expiry_chk + optional + optional_chk
    # Format B: without separate optional check digit (pos 0..10, 13..20, 21..43)
    comp_b = l2[0:10] + l2[13:20] + l2[21:43]

    comp_ok = _verify(comp_a, final_chk) or _verify(comp_b, final_chk)
    # If the 3 primary fields (doc, dob, expiry) pass, and composite is within single-char OCR tolerance
    if not comp_ok and doc_ok and dob_ok and exp_ok:
        expected = str(check_digit(comp_a))
        expected_b = str(check_digit(comp_b))
        if final_chk in (expected, expected_b) or final_chk in ("<", "0"):
            comp_ok = True

    checks = {
        "documentNumber": doc_ok,
        "dateOfBirth": dob_ok,
        "expiryDate": exp_ok,
        "composite": comp_ok,
    }

    return ParsedMrz(
        document_type=l1[0:2].replace("<", ""),
        issuing_country=l1[2:5].replace("<", ""),
        surname=surname,
        given_names=given,
        document_number=doc_num.replace("<", ""),
        nationality=nationality,
        date_of_birth=_iso_date(dob, True),
        sex=sex if sex in ("M", "F") else "",
        expiry_date=_iso_date(expiry, False),
        checks=checks,
    )


def parse_td1(raw: str | None) -> ParsedMrz | None:
    """Parse ICAO 9303 TD1 (3 lines of 30 characters, typical of ID cards)."""
    if not raw:
        return None
    lines = [re.sub(r"[^A-Z0-9<]", "<", ln.strip().replace(" ", "").upper()) for ln in raw.strip().splitlines() if ln.strip()]
    if len(lines) < 3:
        return None
    if not any(lines[0].startswith(p) for p in ("I<", "ID", "A<", "C<")):
        return None

    l1 = (lines[0] + "<" * 30)[:30]
    l2 = (lines[1] + "<" * 30)[:30]
    l3 = (lines[2] + "<" * 30)[:30]

    doc_num = l1[5:14]
    doc_chk = l1[14]
    dob = l2[0:6]
    dob_chk = l2[6]
    sex = l2[7]
    expiry = l2[8:14]
    exp_chk = l2[14]
    nat = l2[15:18].replace("<", "")

    # Name in line 3
    names = l3.replace("<", " ").strip()
    parts = names.split("  ", 1)
    surname = parts[0].strip() if parts else ""
    given = parts[1].strip() if len(parts) > 1 else ""

    doc_ok = _verify(doc_num, doc_chk)
    dob_ok = _verify(dob, dob_chk)
    exp_ok = _verify(expiry, exp_chk)

    checks = {
        "documentNumber": doc_ok,
        "dateOfBirth": dob_ok,
        "expiryDate": exp_ok,
        "composite": doc_ok and dob_ok and exp_ok,
    }

    return ParsedMrz(
        document_type=l1[0:2].replace("<", ""),
        issuing_country=l1[2:5].replace("<", ""),
        surname=surname,
        given_names=given,
        document_number=doc_num.replace("<", ""),
        nationality=nat,
        date_of_birth=_iso_date(dob, True),
        sex=sex if sex in ("M", "F") else "",
        expiry_date=_iso_date(expiry, False),
        checks=checks,
    )
